Fix local redirect policies selecting host networked pods

aditighag opened the pull request:
See commit messages.

Fixes: da35c88eb9f8 ("k8s/watchers: don't silently ignore (*K8sWatcher).updatePodHostData error")
Fixes: 0ab4fa184d3a ("pkg/k8s: ignore certain ipcache errors")
Related: #16920

**Release note**
```release-note
Fix a bug with local redirect policies selecting host networked pods as local endpoints not taking effect.
```

aditighag committed:
bugtool: Collect local redirect policy related data

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
docs: Update kiam example in local-redirect-policy doc

The configuration option is moved from extra args -
https://artifacthub.io/packages/helm/uswitch/kiam/5.9.0.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
k8s/watchers: Fix local redirect policies selecting host networked pods

Commits da35c88eb9f8 and 0ab4fa184d3a introduced a regression for
local-redirect use cases like kiam, whereby host networked pod
updates were skipped. As a result, node-local redirection for
cases where LRPs select host networked pods as backends broke.

Tested the fix on an EKS configured with kiam setup.

Fixes: da35c88eb9f8 ("k8s/watchers: don't silently ignore (*K8sWatcher).updatePodHostData error")
Fixes: 0ab4fa184d3a (pkg/k8s: ignore certain ipcache errors)
Fixes: #16920

Signed-off-by: Aditi Ghag <aditi@cilium.io>

joestringer reviewed (changes requested):
LGTM. One minor nit, the last commit refers to commit 3805bae724e7 but github doesn't recognize it. Is that meant to be 0ab4fa184d3a1a1bf1e7f916bcb5c6e10cf69bb5 ?

joestringer commented on pkg/k8s/watchers/pod.go:
```diff
@@ -665,6 +672,15 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+		logfields.K8sNamespace: newPod.ObjectMeta.Namespace,
+	})
+	if newPod.Spec.HostNetwork {
+		logger.Debug("Pod is using host networking")
```
nit: We don't need to initialize the logger outside the case where it's used. But also I don't anticipate that this is a super hot path, so this shouldn't be too costly. :shrug: Probably fine as-is but worth considering in future if you add new debug statements somewhere that is hit frequently.

aditighag replied:
👍  fixed.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -191,6 +191,13 @@ func (k *K8sWatcher) podsInit(k8sClient kubernetes.Interface, asyncControllers *
 	}
 }
 
+func skipHostNetworkedPod(pod *slim_corev1.Pod) bool {
```
This function name is a little bit misleading because it's not skipping host network pods. What it is doing is checking if it is a hostnetwork pod **and** LRP is disabled.

aanm reviewed (changes requested):
It seems that 2 different commits broke LRP accidentally. I suggest to add e2e tests on this PR that covers the bug this PR is fixing so we don't regress again.

aditighag committed:
test: Add tests for local redirect policy of type address matcher

Test kiam like use cases (LRP select host networked pods) with an address
matcher LRP -
https://docs.cilium.io/en/latest/gettingstarted/local-redirect-policy/#addressmatcher.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
test: Report information to debug failures

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -191,6 +191,13 @@ func (k *K8sWatcher) podsInit(k8sClient kubernetes.Interface, asyncControllers *
 	}
 }
 
+func skipHostNetworkedPodEvents(pod *slim_corev1.Pod) bool {
```
This does not seem addressed from the previous review.

aditighag replied:
@aanm The review feedback has already been addressed as the function is renamed. Also, note that the function returns a bool to indicate if the pod events need to be skipped or not. Since you didn't give any suggestions as to what the function should be renamed to, I used my best judgement.

aanm replied:
That's a fair point. Why not simply remove this function and have the if statement where the function is called? At least that way it is explicit that the event is being skipped because of the `hostNetwork` and the `enableLRP` option. Would that be a fair middle ground?

aditighag replied:
done
